rest-service/manager_rest/app_context.py

```diff
@@ -19,23 +43,66 @@ def get_parser_context(sm=None):
-    raw_parser_context = _extract_parser_context(context)
-    resolver = dsl_parser_utils.create_import_resolver(
-            raw_parser_context['resolver_section'])
-    validate_definitions_version = raw_parser_context[
-        'validate_definitions_version']
-    current_app.parser_context = {
-        'resolver': resolver,
-        'validate_version': validate_definitions_version
-    }
+    current_app.parser_context = _extract_parser_context(context)
 
 
 def _extract_parser_context(context):
     context = context or {}
     cloudify_section = context.get(constants.CLOUDIFY, {})
+    resolver_section = cloudify_section.get(
+        constants.IMPORT_RESOLVER_KEY) or {}
+    resolver_section.setdefault('implementation',
+                                'manager_rest.app_context:ResolverWithPlugins')
+    resolver = dsl_parser_utils.create_import_resolver(resolver_section)
     return {
-        'resolver_section': cloudify_section.get(
-                constants.IMPORT_RESOLVER_KEY),
-        'validate_definitions_version': cloudify_section.get(
-                constants.VALIDATE_DEFINITIONS_VERSION, True)
+        'resolver': resolver,
+        'validate_version': cloudify_section.get(
+            constants.VALIDATE_DEFINITIONS_VERSION, True)
     }
+
+
+class ResolverWithPlugins(DefaultImportResolver):
+    """A resolver which translates plugin-style urls to file:// urls.
+
+    The URL: `file://plugins/cloudify-openstack-plugin/2.0.1` will be
+    translated to: `file:///opt/manager/resources/plugins/<id>`, where <id>
+    is the id of the plugin looked up for the current tenant.
+
+    Both the version and the filename are optional.
+    """
+    PREFIX = 'file://plugins/'
+
+    def fetch_import(self, import_url):
+        if self._is_plugin_url(import_url):
+            resolved = self._resolve_plugin_url(import_url)
+            if resolved:
+                import_url = resolved
+        return super(ResolverWithPlugins, self).fetch_import(import_url)
+
+    def _is_plugin_url(self, import_url):
+        return import_url.startswith(self.PREFIX)
+
+    def _resolve_plugin_url(self, import_url):
+        parts = import_url.replace(self.PREFIX, '').split('/')
+        name = parts[0]
+        version = parts[1] if len(parts) > 1 else None
+        plugin = self._find_plugin(name, version)
+        return self._make_plugin_url(plugin.id)
+
+    def _find_plugin(self, name, version=None):
+        filters = {'package_name': name}
+        if version is not None:
+            filters['package_version'] = version
+        sm = get_storage_manager()
+        return sm.get(Plugin, element_id=None, filters=filters)
+
+    def _make_plugin_url(self, plugin_id):
+        plugin_path = os.path.join(
+            config.instance.file_server_root,
+            FILE_SERVER_PLUGINS_FOLDER,
+            plugin_id)
+        yaml_files = glob.glob(os.path.join(plugin_path, '*.yaml'))
+        if len(yaml_files) != 1:
+            return None
```

Comment:
maybe raising exception?

Reply:
yep now that we changed the `file://` to `plugin:`, it doesn't make sense to return None because there's no way that could be an existing file, thanks